### DISCUSSION TITLE
UserFrosting 5.0 Planning

### DISCUSSION THREAD

**Discussion by lcharette** [userfrosting/UserFrosting#1154] [Ideas & Feature Requests]
Following [my first attempt at moving into the UF 5.0 structure](#1089), I think we need to define a formal plan regarding the repo and class structure for the future version of UserFrosting.

We already know the main goals for v5 are :  
1. Move core, account and admin to their own repo and create a UF skeleton repo / template (#830), so sprinkles are managed by composer, easy to remove if not-needed and easier to test.
1. Add vendor name to sprinkles, to avoid conflicts (#831).
1. Have services providers and bakery commands be deliberately registered instead of auto-discovered

Creating a skeleton repo means all base code not in the core Sprinkle needs to be moved somewhere else, especially the [`sprinklemanager`](https://github.com/userfrosting/UserFrosting/blob/c62790e99e8c68d5f48849bf561a8ed293c0a9f2/app/system/Sprinkle/SprinkleManager.php).

To better illustrate the current situation, here is the current UF repo structure : 
```
- UserFrosting
    - Core
    - Account
    - Admin
- Assets
- Cache
- Config
- Fortress
- I18n
- Session
- Support
- UniformResourceLocator
```

Compared to the proposed structure : 
```
- UserFrosting
- Core Sprinkle
- Account Sprinkle 
- Admin Sprinkle
- Framework
    - Assets
    - Cache
    - Config
    - Fortress
    - I18n
    - Session
    - Support
    - UniformResourceLocator
```

I also generated the current class structure of UF 4 : https://gist.github.com/lcharette/0d6db88d300690468436ed7555cf5ef2#file-1-currentstructure-md

### 1. Where do we draw the line between having classes in the core sprinkle vs. a vendor package?

First, it's important to note we do need a higher lever dependency. For a bare minimum, we need the `SprinkleManager` to be out of core, and ideally out of the skeleton repo too. 

This question is valid for both UF core code and user code. Is there really a requirement for a user or UF extension to create a sprinkle (repo), load the code as a vendor package ? 

I think the answer relies on one thing : **Is the code UserFrosting specific, or can it be reused in another project ?**

If the answer is yes, then I think a vendor package would make sense. But we have to keep in mind all other resources (locale, assets, routes, templates, etc.) still need to be accessible in a specific path (see point 4 about this).

In a functional standpoint, running phpunit on the skeleton repo should run all integration tests related to the content (controller, model, etc.). While all the "unit testing" of the individual component should be done independently. 

I did a quick mockup of what the class/repo structure would be : 
```
- UserFrosting (Skeleton repo)
    - app/
        - cache/
        - logs/
        - sessions/
        - sprinkles/
            - …
        - storage/
        - .env
    - public/
        - index.php
    - vendor/
    - bakery

- UserFrosting Account Sprinkle
    - src/
    - tests/

- UserFrosting Admin Sprinkle
    - src/
    - tests/

- UserFrosting Core Sprinkle
    - src/
    - tests/
 
- UserFrosting-Framework
    - src/
        - Alert
        - Assets
        - Bakery
        - Cache
        - Config
        - Controller
        - Cerf
        - Database
            - Migrator
            - Model
            - Relations
            - Seeder
        - Error
        - Fortress
        - Http
        - I18n
        - Log
        - Mail
        - Router
        - ServicesProvider
        - Session
        - Sprunje
        - Support
        - Sprinkle
        - Tests
        - Throttle
        - Twig
        - UniformResourceLocator
        - Util
    - tests/
```

The full mockup can be found here : 
https://gist.github.com/lcharette/0d6db88d300690468436ed7555cf5ef2#file-3-proposedstructure-md 

Some notes about this structure : 

1. The difference between the **core sprinkle** and the **framework** repo can be seen as  :  
    - All data structure stays in the sprinkle, while all the helper methods moves to the framework repo.
    - This means all Migrations, Model, Seeds, template, locale, assets, etc. stays in the sprinkle. 
1. Some of the `UserFrosting/Framework` content could be split into their own read-only repo, just like Laravel Illuminate packages… But that wouldn’t be required, as the skeleton repo would probably load the whole framework repo.

1. The “framework” namespace wouldn’t be used for simplicity. So it would still be `UserFrosting/Assets/…`, `UserFrosting/i18n/…`, not `UserFrosting/Framework/i18n`. Only sprinkles would be differentiated with the `UserFrosting/Sprinkles/` Namespace. 

1. At this point, I’m not sure how the bakery commands should be handled. Of course, bakery commands should be registered instead of discovered, which means we could put them wherever we want. They could be in the Framework (would make it easier to test) or a sprinkle. Assuming the Bakery class is in the Framework, then I guess it makes sense for command to be found there too. A good example would be the `locale` commands. They don't technically requires UF to run. One could clone the `i18n` repo and want execute those commands. Note [Laravel works that way](https://github.com/illuminate/routing/blob/master/Console/ControllerMakeCommand.php).

1. Same goes for services provider. One thing important about service, currently my IDE doesn’t recognize `$this->ci->translator`  as an instance of `Translator` since it’s actually loaded as a callback inside the provider class. We should implement a custom way to implement something that could take this into account, plus one which is testable.

In the end, I think the key is everything in the `Framework` repo needs to be self contained. They can't have dependencies on the content of the sprinkle (including Models), but it can have dependency on other framework component. 

In other word, **if a component form UF 4 can be moved to a distinct repo right now, it should be moved to the framework repo**.

On the other hand, we also need to take into account dependencies on the locale strings, which would stay inside the sprinkle...

I see the Framework as the bare minimum UF needs to run, but also the one that dictate how it should run. That's why I think this should also apply for **Interfaces**, which should ideally be in the Framework.  

### 2. How do we structure the Sprinkle namespace and directory structure if we add sprinkle vendor name ?

The directory structure would ideally be :

```
    - app/
        - sprinkles/
            - userfrosting/
                - core
                - admin
                - account
            - lcharette/
                - FormGenerator
                - MySite
        - storage
        - sessions
        - …
    - public/
    - vendor/
```

Current namespace for sprinkle is :

- `UserFrosting/Sprinkles/Core`
- `UserFrosting/Sprinkles/Admin`
- `UserFrosting/Sprinkles/FormGenerator`
- `UserFrosting/Sprinkles/MySite`

I see 3 options for the new sprinkle namespaces :

**Option 1 :**
- `UserFrosting/Sprinkles/UserFrosting/Core`
- `UserFrosting/Sprinkles/UserFrosting/Admin`
- `UserFrosting/Sprinkles/Lcharette/FormGenerator`
- `UserFrosting/Sprinkles/Lcharette/MySite`

**Option 2 :**
- `UserFrosting/Sprinkles/Core`
- `UserFrosting/Sprinkles/Admin`
- `Lcharette/Sprinkles/FormGenerator`
- `Lcharette/Sprinkles/MySite`

**Option 3:**
- `Sprinkles/UserFrosting/Core`
- `Sprinkles/UserFrosting/Admin`
- `Sprinkles/Lcharette/FormGenerator`
- `Sprinkles/Lcharette/MySite`

I suggest we stick to the first option, as it’s more “psr-esque”

### 3. Should vendor dir be in `app/vendor/` or `vendor/` ?

I think we should move `vendor/` to the root repo to be consistent with 99% of the PHP world, especially when running command (ie. `vendor/bin/phpunit` vs `app/vendor/bin/phpunit`)

### 4. Should dependent sprinkles be loaded in vendor, and custom user code be left in `app/sprinkles/` ?

For this one, I think we need _some sort_ of control. I've talked about classes so far, but we still need to think about the other resources not in `src/` : assets, locale, routes, config, etc. 

## See also
- [Roadmap for UserFrosting 5](https://github.com/userfrosting/UserFrosting/wiki/Roadmap-for-UserFrosting-5)
- [Silic0nS0ldier/uf5-prototype](https://github.com/Silic0nS0ldier/uf5-prototype)
- [lcharette/uf5-prototype](https://github.com/lcharette/uf5-prototype)
- [lcharette/UserFrosting 5.0](https://gist.github.com/lcharette/0d6db88d300690468436ed7555cf5ef2)

### UF 5.0 Related issue / ideas / wishlist : 
- #1074 
- #1026 
- #1021 
- #836 
- #835 
- #831 
- #830
- #822 
- #813 
- #747 
- #743 
- #729 
- #728 
- #597

**Comment by lcharette** [comment]
One more thing... I'm not sure how to handle the current `Account` sprinkle stuff just yet. I don't see it in the main `framework` repo containing the `Authenticator`. We should be able to left it out of the whole stack if someone doesn't need it, or want to use it's own authenticator. 

This brings to another idea that could be implemented. Instead of dropping everything in the **framework** repo, some feature could also be handled by a new sprinkle. However, this could bring another level of difficulties, in the UI side of things for instance.

For example, a `Notification` feature could either be baked into the Framework, but could also lives inside a sprinkle. My [Breadcrumb](https://github.com/lcharette/UF_Breadcrumb) sprinkle is a good example of this. 

Maybe inside the list of classes / features I listed as potentially be inside the framework, some could be moved to a custom sprinkle, just like `Account` would exist...

**Comment by lcharette** [comment]
Ok I think I have an answer to my own question : 

### 1. Where do we draw the line between having classes in the core sprinkle vs. a vendor package?

With time.

While with time the "framework" repo will probably get bigger and bigger as we move stuff from core to it, for the initial release it should contain the existing support repos plus `System` (main UF class + sprinklemanager) and `Test` : 

```
- UserFrosting-Framework
    - src/
        - Assets
        - Cache
        - Config
        - Fortress
        - I18n
        - Session
        - Support
        - System
        - Tests
        - UniformResourceLocator
    - tests/
```

This will allow the skeleton repo to exist, while allowing it to evolve with time.

As for the existing support repos, we already know having the many repos makes it difficult to develop, tests and release as they are dependent on each other and some needs to be synced to work (eg. [set support to `dev-develop` to test it in other repos, and change it back later](https://github.com/userfrosting/UserFrosting/pull/1070#discussion_r392625797)). So moving them to the framework repo solves this. Plus "SprinkleManager" depends on the locator

I think for 5.0, we can probably leave out sprinkle vendor "for now" and add it in 6.0 for instance.

All while replacing `sprinkles.json` with composer (and `composer,dist`), it will keep the same general class structure for an easier V4 to V5 migration.

So to resume : 
- [ ] Move `UserFrosting/System` namespace to a new repo
- [ ] Move Core, Account & Admin to a separate repo so they can be managed by composer.
- [ ] Remplace `sprinkles.json`, move info to main `composer.json`; Add `composer.json.dist` 
- [ ] Find a better way to register services, one that could be IDE friendly if possible.

**Comment by lcharette** [comment]
Ok guys, @alexweissman @Silic0nS0ldier : Can we at least get the support repo merged in a single Framework one now, and release this for 4.5 ?

Problem is the support packages are currently cross dependent : 

Repo | Depend on
-- | --
Assets | Support
Cache |  
Config | Support
Fortress | i18n
I18n | Support
Session |  
Support | URI
URI |  

1. I18n -> Support -> URI
2. Assets -> Support -> URI
3. Config -> Support -> URI
4. **Fortress -> I18n -> Support -> URI**
5. Support -> URI


I can't fix an issue [in i18n](https://github.com/userfrosting/i18n/runs/1635495205?check_suite_focus=true#step:5:19) without making a fix in `UniformResourceLocator`, but to make sure it's actually resolve the issue, I need to manage `Support` in-between with a temporary commit so it can be picked up by Github Action properly.  

New "Framework" repo would be structure like this : 
```
Framework
  - src/
     - Assets
     - Cache
     - Config
     - Fortress
     - I18n
     - Session
     - Support
     - UniformResourceLocator
  - tests/
     - Assets
     - Cache
     - Config
     - Fortress
     - I18n
     - Session
     - Support
     - UniformResourceLocator
  - composer.json
  - etc.
```

Only issue, we'll lose the git history in the new repo for each independent repo, unless you know a way to conserve it somehow ?

**Comment by Silic0nS0ldier** [comment]
The current dependency graph is a little crazy. 100% it needs attention in UF 5 (or UF 6 if the scope is overloaded), with my thoughts being that `UserFrosting\Support` should be cast into eternal hellfire and package splits either revisited or abolished in favour of a monolithic library.

However this goes, switching to a project based structure (where 90% or more UF code is brought in by composer) should help quite a bit.

Oh, and this comment only relates to https://github.com/userfrosting/UserFrosting/issues/1095#issuecomment-753737063, I haven't read the rest of this thread.

**Comment by lcharette** [comment]
(Looks like preserving the git history _could_ be done : https://medium.com/@checko/merging-two-git-repositories-into-one-preserving-the-git-history-4e20d3fafa4e)

**Comment by lcharette** [comment]
_Moved this issue to discussion as it's look like a more appropriate place to discuss this_


We're now a year later from the original post, and after a quick detour working with Laravel, I have a clear view of how UF should be structured. Funny thing is, 90% of what I wrote here a year ago still reflect my current plan. What changed are : 

> 2. How do we structure the Sprinkle namespace and directory structure if we add sprinkle vendor name ?
> 4. Should dependent sprinkles be loaded in vendor, and custom user code be left in app/sprinkles/ ?

I think the answer is between option 2 and no options at all. As long as the sprinkles services and assets are registered, it doesn't matter what namespace they use.  I don't think it will matter anymore where the sprinkle are loaded. It makes 100% sense to me to have then in _vendor_ and have the "project" code in `app/` _à la Laravel_. But, for simpler stuff, we might keep the `app/sprinkles` dir for manual installation (for newbs), I don't know... As long as everything is explicitly registered in a PHP class compared to autodetected with voodoo magic applied to a specific namespace or path convention, it doesn't really matter anymore.


So the current short term plan to get this started would be : 

 ### UF 4.6
- [x] Create [`userfrosting/framework`](https://github.com/userfrosting/framework)
- [x] Upgrade to Laravel 8.x and PHP 8; Drop PHP 7.2 & 7.4 support (#1074)

### UF 5
- [ ] Move UserFrosting/System namespace, and all possible code out of main repo and to `userfrosting/framework`
- [ ] Move Core, Account & Admin to a separate repo so they can be managed by composer (#830), effectively creating the UF skeleton repo;
- [ ] Locale separation (#1026) 
- [ ] Remove `sprinkles.json`, move info to main `composer.json.dist` (#813; #747; #729)
- [ ] Handling of Sprinkle Vendor name (#831);
- [ ] Find a better way to register Sprinkle services and assets (#1021; #743)
- [ ] Potentially create a new modular experience and UI (eg. `admin-html` & `admin-Vue` UI you can choose from) (#836);


Meanwhile, I started working on the framework repo yesterday : https://github.com/userfrosting/framework

**Reply by Silic0nS0ldier** [reply]
Been experimenting and ideating on Sprinkle registration over at https://github.com/Silic0nS0ldier/uf5-prototype. Got a functional (if not fully tested, circular dependency handling tests are to be implemented) mechanism for dependency registration where the load ordering is automatically worked out.

I'm not thrilled with it, currently trying to work out a solution that avoids string based reflection entirely. But at least it (seems) to work and importantly doesn't do something cringe-worthy like dynamically generating the fully qualified class name.

**Reply by lcharette** [reply]
One thing for sure, it need a lots more of those Interface 😅. I've also been reading about PSR-14, might be helpful too.

**Reply by lcharette** [reply]
As discussed in Chat, I edited the previous post. The first part will be as 4.6 and 5.0 (instead of 5 and 6)

> ### UF 4.6
> ...
> 
> ### UF 5
> ...

**Reply by lcharette** [reply]
(also, "Move UserFrosting/System namespace to a new repo" is moved to UF5)

**Reply by lcharette** [reply]
Note, since UF 4.6 is a thing now, I think it would be best for UF 5 to be PHP 8 and up only. By the time we ship it, PHP 7.4 might be out of support. 4.6 could be out LTS version.

**Reply by Silic0nS0ldier** [reply]
Making PHP 8 the minimum also lets us leverage a lot of the new PHP 8 features (notably attributes, union types and null safe operators).

**Comment by Silic0nS0ldier** [comment]
Splitting the UI out in UF5 means we'll want to make some common UI components to help developers build out their sites more quickly. (which I think I previously mentioned).

There is a tool that will help a lot on this front, StorybookJS. It helps a lot with building UI components by providing an environment where they can be iterated on in isolation. It's also handy for implementing visual regression tests, and documenting component usage.

It's something to look into including in whatever UI starter project we land on as well.

**Reply by Silic0nS0ldier** [reply]
Had a go of adding Storybook to some web components. Didn't go so well... Build wise its quite opinionated by default which made integrating hard (not helped by it being behind on upgrading webpack).

That said, I work a workable solution. Turns out it plugged in quite nicely with the React package, so its living there for now. Long term I'll probably split it into its own isolated folder. Should separate the build processes, making for less pain and easy updates.

https://github.com/Silic0nS0ldier/uf5-prototype/tree/master/lib-gui/ui-base/react-components/src/stories

**Comment by Silic0nS0ldier** [comment]
To support custom frontends (server composed HTML, Vue, React, Angular, Svelte, etc), we need a way to provide components that are compatible across most ecosystems. Web components seem the best choice (individual frontend types may need adapters, however they will be minimal), provided we have a library helping with construction.

A few libraries of note are;

* https://hybrids.js.org/ 
  Very simple API surface, appears easy to test components, a little heavy clocking in around 10.1kB (minified and gzipped)
* https://lit.dev/ 
  Backed by Google, reliant on experimental decorators, strong ecosystem, 5.8kB (minified and gzipped)
* https://www.fast.design/ 
  Backed by Microsoft, optionally uses experimental decorators, feature dense including design system APIs and state management, closer to a framework then a library, little heavy at 11.1kB (minified and gzipped)
* https://stenciljs.com/ 
  React-like component composition, able to create adapters for other ecosystems out-of-the-box, strong async support good fit for data-centric areas in UF, 6.8kB (minified and gzipped) but likely smaller in practise (part of lib is a compiler, is tree-shakeable), personal favourite

Questions that need answers;

1. What library do we want to use for component creation?
2. Do we want to only support Vue and HTML initially to keep scope reined in?
3. What will the HTML usage look like? How will only the needed components be loaded on a given page?
4. What will be the minimum browser version? Ideally we will only work with ES module code.
5. What components do we need/want?
6. What existing components depend on third party libraries? Which of them can be reused, and do we want to reuse them?

**Reply by lcharette** [reply]
I'll leave frontend question to you as your look more experienced than me on that point (I've only used vuejs for less than a year now, and don't know any of the link you shared). 

I'll only point that our user base probably only ever care about the theming (what it looks like) : Should we use Bootstrap, AdminLTE, CoreUI...? FYI, I'm open minded to rework the whole UI for V5 from start if it need to (so drop AdminLTE if necessary). This is one of Laravel shortcoming IMO, especially Laravel Jetstream : You're not only tied to either Vue/Blade, but also Tailwind CSS. 

But since V5 is meant to be modular, I see `admin-vue` and `admin-html` sprinkles, and end user chose which one he want to works with. We could even go as far as `admin-vue-boostrap` and `admin-vue-adminlte`. We could probably do same thing with "core" / non admin pages. My only concern is one of those might be out of sync with the other feature wise.

**Reply by lcharette** [reply]
> 2. Do we want to only support Vue and HTML initially to keep scope reined in?

I do think so. On my site, I'm 100% on cue right now. But it makes sense to have both.

**Reply by lcharette** [reply]
(I'm gonna need to read more about Web Components...)

**Reply by Silic0nS0ldier** [reply]
> This is one of Laravel shortcoming IMO, especially Laravel Jetstream : You're not only tied to either Vue/Blade, but also Tailwind CSS.

If everything goes to plan, the final implementation should be quite flexible. Business logic and the UI will be decoupled, which should make it easy for users to theme to their hearts delight. (be it changing colours or completely changing how something is laid out)

**Reply by Silic0nS0ldier** [reply]
I've started trying out StencilJS, so far things are looking good.

https://github.com/Silic0nS0ldier/uf5-prototype/tree/master/lib-gui/ui-base

**React**
React has abysmal support for web components out of the box, requiring a bunch of boilerplate to get it properly working. StencilJS has tooling to do this for us thankfully, and it works really well. Usage wise, consumers import React components named after the web component that powers them behind the scenes. Best part is that TypeScript types are carried over so concerns should have a really good dev experience.

A dedicated package will be required.

**Preact**
Unlike React, Preact has _really_ good support for web components. To start using them it is only necessary to call the define custom elements function. TypeScript types don't appear to be automatic (citing the explicit function call being required), but should be easy to enable.

**Vue**
When Vue comes across a HTML tag that is non-standard, it assumes this must be a Vue component. If no matching Vue component is found, an error is thrown (reportedly anyway). There is an API to expand the browser handled elements, which along with calling the define elements function is all that is needed to support web components.

**HTML**
Nothing particularly special here, since this is the common denominator for everything. Call the define elements function and then get to work.

**Comment by Silic0nS0ldier** [comment]
Something to think about with the splitting of the frontend in v5 is that in a system with authentication, there is often a need to serve static assets/content that is not available to all. This isn't necessarily an area to be addressed in the initial v5 release but is something worth keeping in mind. https://github.com/userfrosting/UserFrosting/issues/1161

**Comment by lcharette** [comment]
I've been pushing code to different part of V5 for some time now and wanted to take a moment share current status of what I'm doing. I also need to poke @Silic0nS0ldier thought on Assets.

Parts that have been worked on for now : 
- [userfrosting/framework@develop-5.0](https://github.com/userfrosting/framework/tree/develop-5.0)
- [userfrosting/sprinkle-core@5.0](https://github.com/userfrosting/sprinkle-core)
- [userfrosting/UserFrosting@develop-5.0](https://github.com/userfrosting/UserFrosting/tree/develop-5.0)

See also my todo (might not be public) : https://github.com/orgs/userfrosting/projects/5

## Overall Structure
The current overall structure of V5 works around the skeleton (main) repo. This one is tasked with handling composer dependencies which now includes all sprinkles except the base app.

### Dependencies and requirements
As decided early on, UF5 requires PHP 8 or (eventually) earlier. I'm already using some new PHP 8 features like named arguments, constructor property promotion and union type. It's great !

UF5 also runs on **Slim4** and uses [**PHP-DI 6**](https://php-di.org) as the new dependency injection container. PHP-DI is **TOTALLY AWESOME** and while the very nature of services providers will change from V4 to adapt to PHP-DI features, this will be for the best. 

Classmapper won't be necessary anymore as this duty can be handled by PHP-DI directly. Autowiring and Annotation injection makes it easy to inject services and other dependencies into pretty much anything, including Symfony Console command (Bakery commands).

And I've been able to optimized most of the PHPUnit tests so far, and more optimization will comes from better use of dependencies injections.

### Main UserFrosting Repo
This repo is the skeleton every UF5 app will start from. It will (eventually) host the default (working) website end users will modify. Goal here is for this skeleton to be pushed back to end users own repo and requires minimal maintenance when it's time for updates. I plan on making sure all files included in this skeleton are clearly identify as to what can be edited and deleted as necessary (unlike Laravel...).

For example, if I want to create "MyApp" using UF5, I would clone the main UF repo and put everything related to "MyApp" in there. Here's the default Skeleton Sprinkle Recipe as an example: https://github.com/userfrosting/UserFrosting/blob/96f95f60b86935e048f6798ecb43ffbba86b692c/app/src/MyApp.php. 

As of today, maybe 20% of this repo is done.

### Framework 
The Framework repo holds the legacy support packages (Asssets, config, i18n, etc.) for now, plus the backbone requires to launch a modular *Slim* (web) app and Symfony Console (Bakery) app. 

I've made it so each Sprinkle now has a [`SprinkleRecipe`](https://github.com/userfrosting/framework/blob/develop-5.0/src/Sprinkle/SprinkleReceipe.php) class. This class is used to register the sprinkles Bakery Commands, Routes, Services and Middlewares (and eventually migrations) with the main Slim/Console app. Each sprinkle can also define other `SprinkleRecipe` as dependent. When the app is initialized, the sprinkle tree will be compiled and everything registered in the correct order.

For example, here's the *Core Sprinkle* recipe : https://github.com/userfrosting/sprinkle-core/blob/dce40cc0c7de09983e74ad2bcbe89988152c4039/src/Core.php

This means all sprinkles (from UF or 3rd party) are now handles exclusively by Composer and every feature is now registered instead of magically added based on namespace convention or path locations. 

A higher priority sprinkle could also extend another Sprinkle. Only want parts of `admin` sprinkle? Include it with composer, but instead of registering the whole Sprinkle, register only the parts you needs in your own `MyApp` Sprinkle Recipe.

Note that while a "dependent sprinkle" (that doesn't run on it's own and doesn't contain a `public/` directory for example) doesn't requires to be made from the skeleton, each and every UF5 app created using the skeleton can become a "dependent sprinkle". 

It also means a sprinkle could be located **anywhere** now. It can be in the `vendor/` directory or another sprinkle, as long as it's a PSR-4 compliant namespace.

It's important to note I made it so the Framework has very minimal code, as the rest is still located in the *Core Sprinkle*. There's couple reason for this:

1. The Sprinkle Manager cannot include itself (so can't be in Core);
2. Framework has minimal services it can register, otherwise it would need to include itself (see #1). So as many services and Routes and other consumables as possible need to be in a Sprinkle.
3. You _could_ remove the whole *Core Sprinkle* and get a very basic Slim 4 / Symfony Console app (without db, assets, i18n, etc.). Only Slim, PHP-DI, CLI Commands, Routes and Middlewares... and the Sprinkle Manager.

At this point, parts of framework _could_ still be moved to Core or _vice versa_, and things could probably be renamed as well. We could completely go nuclear on this and makes a lot of things their own repo/sprinkle (Don't need db for your app? Don't install the sprinkle !) but I think for now, for the amount of market share UF has and the dev team it has (me), it not worth it. But it still been design to ! 

Anyhow, I consider the Framework to be 90% complete.

### Core Sprinkle 
Most has been said in the Framework part, but one key difference with the new Core Sprinkle aside from being it's own repo, the "Default" pages won't be included in it anymore. Those are moved to the Skeleton. In fact, Core will only include the assets and alerts routes so far. 

Core sprinkle should also include the necessary services and whatnot to enable Twig 3, i18n, database, config, migrations, etc. support.

I'm still working on integrating all the parts of the Core sprinkle with PHP-DI and the new structure. Current task is upgrading all existing Unit tests so everything gets green and gets to a minimal working state before going further. 

### Admin & Account sprinkles
No work as been done on that part yet. But the key point here is `Admin` will be able to depend on `Account`. So while building you app, you won't require to explicitly requires `Account` if you requires `Admin` (in Composer, as well as in the Sprinkle Recipes).

## Assets
That's the moment I ping @Silic0nS0ldier again, and the whole reason for this post.

With the new structure, **Assets management** should also be moved to Core **or** Framework. Minimal maintenance should be required from the files included in the skeleton. This means much of the the Node/NPM files in the main repo needs to move.

I've made some attempts to get assets working (the Bakery side is fine), but the node/npm side will require much work. Thing is, it's been initially design with `app/sprinkles.json` file in mind, which I nuked for PHP classes. It was also initially designed to find all `packages.json` inside `sprinkles/*/`, which won't be the case.

I see two solutions to this problem, but there could be other.

1. Have PHP give (or generate) the required information to Node
2. Have a separate definition file for assets

For #1, I think it might be overkill to have PHP run some tasks currently done by Node, but at the same time while the frontend command can be run on their own, they've typically been called using the Bakery command (PHP) as a middleman. So it's not _too farfetched_ to think Bakery could first compile a json file containing the required info for Node to do it's thing.

Solution #2 might be simpler, but It think it will create major overlaps. Since Composer/PSR-4 manage the location of file on the PHP side, an equivalent would be needed on the JS side, but also an equivalent of the Sprinkle Manager. I also think if you include a sprinkle in composer/Sprinkle Recipe, it's assets should be included. So I see this solution more of a temporary solution while we work on something else.

I've yet to investigate how Symfony/Laravel/Slim handle this for their own "packages" or "plugins", but I know we won't find the ultimate solution on the Laravel side...

**Reply by Silic0nS0ldier** [reply]
Re the frontend, not much progress has been made (competing priorities). That is not however to say no progress has been made.

## Web Components

As previously mentioned, web components will be utilised to avoid framework lock-in. An early exploration has commenced, using [Stencil](https://stenciljs.com/) to permit generation of adapters for popular frameworks including;

* React, which has major incompatibilities with web components which are resolved via the adapter.
* Vue, which has great support but needs to be instructed on what native components exist.

Other frameworks (e.g. Preact) and "native web" (e.g. UF v4 and earlier) have no specific requirements, web components can be used provided they have been registered.

## Design System with Atomic Design

[Atomic Design](https://atomicdesign.bradfrost.com/) details an approach for creating and maintaining robust design systems. A key takeaway for me is the ["Atomic Design Methodology"](https://atomicdesign.bradfrost.com/chapter-2/) where UI components are decomposed into the following;

1. Atoms
2. Molecules
3. Organisms
4. Templates
5. Pages

My goal is to take this approach to UI decomposition, and apply it to the in-box UserFrosting UI. Before that however I'm applying it to a smaller scoped project that I hope will mature into a useful tool.

## Monolog Explorer

[Monolog Explorer](https://github.com/Silic0nS0ldier/monolog-explorer) is an effort to create a small-scale monolog log file processor suited to young websites and local development. It will consist of 2 parts;

1. A web app (responsible for presentation and processing of log file data)
2. A relay API (responsible for providing access to server-side logs, likely to be delayed to a phase 2)

Relating back to the design system, this project will act as a testbed for the following;

* Application of the Atomic Design Methodology
* Theming via CSS variables (with light/dark theming)
* Assessing viability using of CSS's `@import` rule as the basic for CSS bundling (if things go well then a smarter CSS-JS integration may be possible, though this is in the post UF5 timeframe)
* Assessing [Modern Web](https://modern-web.dev/) web dev tool suite
* Creating installable web apps (PWA)
* Creating store distributable, natively enhanced apps (HWA)
